iwyu_ast_util.cc

```diff
@@ -588,17 +594,37 @@ class TypeEnumerator : public RecursiveASTVisitor<TypeEnumerator> {
     // TypedefForA would be inserted in seen_types_ for Tpl<A, TypedefForA>,
     // leading to indeterminate resugar_map construction.
     // TODO(bolshakov): check whether typedef components are "provided".
-    return Base::TraverseType(QualType(desugared, 0));
+    return TypeEnumerator<TemplateArgEnumerator>::Base::TraverseType(
+        QualType(desugared, 0));
   }
+};
 
-  set<const Type*> seen_types_;
+class TypeEnumeratorWithoutSubstituted
+    : public TypeEnumerator<TypeEnumeratorWithoutSubstituted> {
+ public:
+  bool TraverseTypeHelper(QualType qual_type) {
+    CHECK_(!qual_type.isNull());
+    const Type* type = qual_type.getTypePtr();
+    if (type->getAs<SubstTemplateTypeParmType>())
+      return true;
+
+    seen_types_.insert(GetCanonicalType(type));
+
+    return TypeEnumerator<TypeEnumeratorWithoutSubstituted>::Base::TraverseType(
```

Comment:
Let's add a `typedef TypeEnumerator<TypeEnumeratorWithoutSubstituted> Base;` in this class, and use `retur Base::Base::TraverseType(...);`.

The double-Base looks a little funny, but it accurately highlights how far up the lineage we're jumping.

Reply:
And same in the other derived TypeEnumerator.